scope.go

```diff
@@ -0,0 +1,262 @@
+	for s := s; s != nil; s = s.parentScope {
+		stores = append(stores, s)
+	}
+	for i, j := 0, len(stores)-1; i < j; i, j = i+1, j-1 {
+		stores[i], stores[j] = stores[j], stores[i]
+	}
+	return stores
+}
+
+func (s *Scope) knownTypes() []reflect.Type {
+	typeSet := make(map[reflect.Type]struct{}, len(s.providers))
+	for k := range s.providers {
+		typeSet[k.t] = struct{}{}
+	}
+
+	types := make([]reflect.Type, 0, len(typeSet))
+	for t := range typeSet {
+		types = append(types, t)
+	}
+	sort.Sort(byTypeName(types))
+	return types
+}
+
+func (s *Scope) getValue(name string, t reflect.Type) (v reflect.Value, ok bool) {
+	v, ok = s.values[key{name: name, t: t}]
+	return
+}
+
+func (s *Scope) setValue(name string, t reflect.Type, v reflect.Value) {
+	s.values[key{name: name, t: t}] = v
+}
+
+func (s *Scope) getValueGroup(name string, t reflect.Type) []reflect.Value {
+	items := s.groups[key{group: name, t: t}]
+	// shuffle the list so users don't rely on the ordering of grouped values
+	return shuffledCopy(s.rand, items)
+}
+
+func (s *Scope) submitGroupedValue(name string, t reflect.Type, v reflect.Value) {
+	k := key{group: name, t: t}
+	s.groups[k] = append(s.groups[k], v)
+}
+
+func (s *Scope) getValueProviders(name string, t reflect.Type) []provider {
+	return s.getProviders(key{name: name, t: t})
+}
+
+func (s *Scope) getGroupProviders(name string, t reflect.Type) []provider {
+	return s.getProviders(key{group: name, t: t})
+}
+
+func (s *Scope) getProviders(k key) []provider {
+	nodes := s.providers[k]
+	providers := make([]provider, len(nodes))
+	for i, n := range nodes {
+		providers[i] = n
+	}
+	return providers
+}
+
+func (s *Scope) getAllGroupProviders(name string, t reflect.Type) []provider {
+	return s.getAllProviders(key{group: name, t: t})
+}
+
+func (s *Scope) getAllValueProviders(name string, t reflect.Type) []provider {
+	return s.getAllProviders(key{name: name, t: t})
+}
+
+func (s *Scope) getAllProviders(k key) []provider {
+	allScopes := s.getScopesFromRoot()
+	var providers []provider
+	for _, scope := range allScopes {
+		providers = append(providers, scope.getProviders(k)...)
+	}
+	return providers
+}
+
+func (s *Scope) invoker() invokerFn {
+	return s.invokerFn
+}
+
+// adds a new graphNode to this Scope and all of its descendent
+// scope.
+func (s *Scope) newGraphNode(wrapped interface{}, orders map[*Scope]int) {
+	orders[s] = s.gh.NewNode(wrapped)
+	for _, cs := range s.childScopes {
+		cs.newGraphNode(wrapped, orders)
+	}
+}
+
+func (s *Scope) cycleDetectedError(cycle []int) error {
+	var path []cycleErrPathEntry
+	for _, n := range cycle {
+		if n, ok := s.gh.Lookup(n).(*constructorNode); ok {
+			path = append(path, cycleErrPathEntry{
+				Key: key{
+					t: n.CType(),
+				},
+				Func: n.Location(),
+			})
+		}
+	}
+	return errCycleDetected{Path: path, scope: s}
+}
+
+// String representation of the entire Scope
+func (s *Scope) String() string {
+	b := &bytes.Buffer{}
+	fmt.Fprintln(b, "nodes: {")
```

Comment:
yea it's a valid point, though we're writing it to a bytes.Buffer and not some external stream, and doing err check for every single Fprintln in this func can be expensive + hurts readability. i can add it though if you think the tradeoff is worth.

Reply:
nah, i think you are just adding existing code. just thought i would mention in case it comes up again.

Reply:
FWIW bytes.Buffer.Write [always returns a nil error](https://pkg.go.dev/bytes#Buffer.Write), so checking errors here would be unnecessary. I don't personally use errcheck much, but it will also ignore this case when the writer is known to be a bytes.Buffer at compile time.